coderd/gitauth/config_test.go

```diff
@@ -3,36 +3,93 @@ package gitauth_test
 	"golang.org/x/xerrors"
 
+	"github.com/coder/coder/v2/coderd"
+	"github.com/coder/coder/v2/coderd/coderdtest/oidctest"
 	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
 	"github.com/coder/coder/v2/coderd/database/dbfake"
-	"github.com/coder/coder/v2/coderd/database/dbgen"
 	"github.com/coder/coder/v2/coderd/gitauth"
 	"github.com/coder/coder/v2/codersdk"
 	"github.com/coder/coder/v2/testutil"
 )
 
 func TestRefreshToken(t *testing.T) {
 	t.Parallel()
-	t.Run("FalseIfNoRefresh", func(t *testing.T) {
+	const providerID = "test-idp"
+	expired := time.Now().Add(time.Hour * -1)
+
+	t.Run("NoRefreshExpired", func(t *testing.T) {
 		t.Parallel()
-		config := &gitauth.Config{
-			NoRefresh: true,
-		}
-		_, refreshed, err := config.RefreshToken(context.Background(), nil, database.GitAuthLink{
-			OAuthExpiry: time.Time{},
+		fake, config, link := setupOauth2Test(t, testConfig{
+			FakeIDPOpts: []oidctest.FakeIDPOpt{
+				oidctest.WithRefreshHook(func(_ string) error {
+					t.Error("refresh on the IDP was called, but NoRefresh was set")
+					return xerrors.New("should not be called")
+				}),
+				// The IDP should not be contacted since the token is expired. An expired
+				// token with 'NoRefresh' should early abort.
+				oidctest.WithDynamicUserInfo(func(_ string) (jwt.MapClaims, error) {
+					t.Error("token was validated, but it was expired and this should never have happened.")
+					return nil, xerrors.New("should not be called")
+				}),
+			},
+			GitConfigOpt: func(cfg *gitauth.Config) {
+				cfg.NoRefresh = true
+			},
 		})
+
+		ctx := oidc.ClientContext(context.Background(), fake.HTTPClient(nil))
+		// Expire the link
+		link.OAuthExpiry = expired
+
+		_, refreshed, err := config.RefreshToken(ctx, nil, link)
 		require.NoError(t, err)
 		require.False(t, refreshed)
 	})
+
+	// NoRefreshNoExpiry tests that an oauth token without an expiry is always valid.
+	// The "validate url" should be hit, but the refresh endpoint should not.
+	t.Run("NoRefreshNoExpiry", func(t *testing.T) {
+		t.Parallel()
+
+		validated := false
+		fake, config, link := setupOauth2Test(t, testConfig{
+			FakeIDPOpts: []oidctest.FakeIDPOpt{
+				oidctest.WithRefreshHook(func(_ string) error {
+					t.Error("refresh on the IDP was called, but NoRefresh was set")
+					return xerrors.New("should not be called")
+				}),
+				oidctest.WithDynamicUserInfo(func(_ string) (jwt.MapClaims, error) {
+					validated = true
+					return jwt.MapClaims{}, nil
+				}),
+			},
+			GitConfigOpt: func(cfg *gitauth.Config) {
+				cfg.NoRefresh = true
+			},
+		})
+
+		ctx := oidc.ClientContext(context.Background(), fake.HTTPClient(nil))
+
+		// Zero time used
+		link.OAuthExpiry = time.Time{}
+		_, refreshed, err := config.RefreshToken(ctx, nil, link)
+		require.NoError(t, err)
+		require.True(t, refreshed, "token without expiry is always valid")
+		require.True(t, validated, "token should have been validated")
+	})
+
```

Comment:
This was one test, I made it two. Testing both an expired and and a `IsZero` time.